kernel/src/grant.rs

```diff
@@ -488,33 +836,73 @@ impl<'a, T: Default> ProcessGrant<'a, T> {
             })
             .ok();
 
-        // Return early if no grant
-        let grant_ptr = if let Some(ptr) = optional_grant_ptr {
+        // Return early if no grant. Type annotation for unsafe correctness.
+        let grant_ptr: *mut u8 = if let Some(ptr) = optional_grant_ptr {
             ptr
         } else {
             return None;
         };
 
+        // See new() for more explanation on these calculations.
+        let upcalls_size = size_of::<usize>() + (NUM_UPCALLS * size_of::<SavedUpcall>());
+        let grant_t_align = align_of::<T>();
+        let upcalls_padding = grant_t_align - (upcalls_size % grant_t_align);
+
+        // `grant_ptr` now refers to the special memory we store for each
+        // `Grant` which contains the number of upcalls for this grant, an array
+        // of upcall data, some potential padding, and then the object of type
+        // T.
+        //
+        // To get to the correct pointer where object of type T is store, we
+        // have to increment the pointer past our saved upcall state and
+        // padding.
+        //
+        // # Safety
+        //
+        // This pointer is safe because it is a *u8 and we offset it the same
+        // number of bytes as when the grant memory was originally allocated.
+        let grant_type_ptr = unsafe { grant_ptr.add(upcalls_size + upcalls_padding) };
+
+        // # Safety
+        //
+        // This pointer is safe because it is a *u8 and we offset it the correct
+        // number of bytes to the start of the `SavedUpcall` array.
+        let saved_upcalls_ptr = unsafe { grant_ptr.add(size_of::<usize>()) };
+        // # Safety
+        //
+        // Creating this slice is safe because the pointer is in the grant
+        // allocation that is guaranteed to still exist, a grant can only be
+        // entered once at a time so this is the only mutable reference to this
+        // slice, and the slice has valid SavedUpcalls which are guaranteed to
+        // be initialized.
+        let saved_upcalls_slice = unsafe {
+            slice::from_raw_parts_mut(saved_upcalls_ptr as *mut SavedUpcall, NUM_UPCALLS)
```

Comment:
Can this be a `slice::from_raw_parts_mut` here?

Reply:
I'm not sure what you mean -- it already is a `slice::from_raw_parts_mut` here? Is the concern that using it this way is unsafe?

Reply:
Ah, sorry! Should've been "can this be a `slice::from_raw_parts`" (i.e. non-mutable) here? Not a safety concern per se.